rasa/nlu/classifiers/diet_classifier.py

```diff
@@ -590,63 +515,149 @@ def _compute_default_label_features(
+                    f"sequence level features for {self.label_attribute}. "
+                    f"Falling back to using default one-hot embedding vectors. "
+                )
             sequence_features = None
-            sentence_features = self._compute_default_label_features(labels_example)
+            sentence_features = self._compute_default_label_features(
+                num_labels=len(label_id_dict)
+            )
 
+        # Save them as a RasaModelData
         label_data = RasaModelData()
         label_data.add_features(LABEL, SEQUENCE, sequence_features)
         label_data.add_features(LABEL, SENTENCE, sentence_features)
-        if label_data.does_feature_not_exist(
-            LABEL, SENTENCE
-        ) and label_data.does_feature_not_exist(LABEL, SEQUENCE):
-            raise ValueError(
-                "No label features are present. Please check your configuration file."
-            )
 
-        label_ids = np.array([idx for (idx, _) in labels_idx_examples])
-        # explicitly add last dimension to label_ids
-        # to track correctly dynamic sequences
+        # Add label id features - and sequence length information (if sequence
+        # features are present)
         label_data.add_features(
             LABEL_KEY,
             LABEL_SUB_KEY,
-            [FeatureArray(np.expand_dims(label_ids, -1), number_of_dimensions=2)],
+            [FeatureArray(np.expand_dims(label_indices, -1), number_of_dimensions=2)],
         )
-
         label_data.add_lengths(LABEL, SEQUENCE_LENGTH, LABEL, SEQUENCE)
 
         return label_data
 
+    def _collect_one_example_per_label(
+        self, messages: List[Message], label_id_dict: Dict[Text, int],
+    ) -> Tuple[np.ndarray, List[Message]]:
+        """Collect one example message for each label.
+
+        Args:
+          messages: the messages which
+
+        Returns:
+          a sorted array containing the indices from the given `label_id_dict`
+          exactly and a list of messages where the i-th message is an example for
+          the label whose index is given by the i-th entry in the array
+        """
+        labelidx_message_tuples = []
+        debug_unsupported_labels = set()
+        for label_name, label_idx in label_id_dict.items():
+            message = next(
+                (
+                    message
+                    for message in messages
+                    if message.get(self.label_attribute) == label_name
+                ),
+                None,
+            )
+            if message is None:
+                debug_unsupported_labels.update([label_name])
+            labelidx_message_tuples.append((label_idx, message))
+
+        if len(debug_unsupported_labels):
+            raise InvalidConfigException(
+                "Expected at least one example for each label "
+                "in the given training_data but could not find "
+                f"any example(s) for {sorted(debug_unsupported_labels)}."
+            )
```

Comment:
You can also pass it in as a parameter during initialisation. .... And this smells like another potential bug:
- you can pass in a self.index_label_id_mapping
- but that self.index_label_id_mapping is overwritten when you preprocess_train_data 
Let's say we want to finetune on a small dataset where some labels do not appear. That looks like it would lead to problems, no? 
I suggest that we: use the mapping passed via the constructor, only overwrite it in case it's None, and keep the check and either leave the exception or just raise a warning (?)

Reply:
> You can also pass it in as a parameter during initialisation

initialization of what? `DIETClassifier`?

> Let's say we want to finetune on a small dataset where some labels do not appear. That looks like it would lead to problems, no?

Finetuning inside Rasa Open Source always requires access to old data as well and there are checks before `DIETClassifier` to ensure that a `TrainingData` object with incomplete number of labels is not passed to DIET.

> I suggest that we: use the mapping passed via the constructor, only overwrite it in case it's None

I'm assuming you are talking about `label_id_dict`? It will only be `None` when the model is trained for the first time from scratch. In the case when a trained model is loaded, it will always be passed from the `load` method irrespective of whether fine-tuning or not. During fine-tuning, `preprocess_train_data` will "try" to overwrite it but it is guaranteed to give the same mapping because of the finetuning argument I mentioned above.

Reply:
> Finetuning inside Rasa Open Source always requires access to old data as well
I'll put it on my todo list to understand why. Might ask you some more questions on this later ^^

> During fine-tuning, preprocess_train_data will "try" to overwrite it but it is guaranteed to give the same mapping because of the finetuning argument I mentioned above.
so "try  to overwrite" means it "does overwrite it with the same data (because of things happening in other classes)"

Since this check doesn't cost us anything - do we keep it or shall I remove it?

Reply:
should be easy to add, I will try

Reply:
added test